drv/lpc55-swd/src/main.rs

```diff
@@ -0,0 +1,656 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use drv_lpc55_spi as spi_core;
+use drv_lpc55_syscon_api::{Peripheral, Syscon};
+use drv_sp_ctrl_api::SpCtrlError;
+use idol_runtime::{
+    LeaseBufReader, LeaseBufWriter, Leased, LenLimit, RequestError, R, W,
+};
+use lpc55_pac as device;
+use ringbuf::*;
+use userlib::*;
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    Idcode(u32),
+    Idr(u32),
+    MemVal(u32),
+    SwdRead(Port, RawSwdReg),
+    SwdWrite(Port, RawSwdReg, u32),
+    ReadCmd,
+    WriteCmd,
+    None,
+    AckErr(Ack),
+}
+
+ringbuf!(Trace, 64, Trace::None);
+
+task_slot!(SYSCON, syscon_driver);
+task_slot!(GPIO, gpio_driver);
+
+#[derive(Copy, Clone, PartialEq)]
+enum Ack {
+    //Ok,
+    Wait,
+    Fault,
+    Protocol,
+}
+
+// Several of these bits in the CSW are  implementation defined according to
+// the official specification but thankfully RM0433 60.4.2 actually gives us the
+// defintions
```

Comment:
You reference the ARM ADI docs below -- they're the authoritative definition of the CSW contents for an AHB MEM-AP, not the ST RM, fwiw.